### ISSUE TITLE
Delay memtx snapshot until quorum for the op is achieved

### ISSUE THREAD

**Issue by sergos** [tarantool/tarantool#4850]
The [qsync](https://github.com/tarantool/tarantool/blob/sergos/quorum-based-synchro/doc/rfc/quorum-based-synchro.md) replication relies on postponed confirmation of txn through the special quorum operation. 
Existent memtx snapshot mechanism should be modified to wait for a quorum message to arrive before the actual snapshot is performed. So, the readview should be created off some txn and snapshot should start only after quorum message for this lsn (or a bigger one) is received.

In case master connection is lost the snapshot should be performed in the following way:
- rollback all transactions that are waiting for quorum confirmation - this brings consistent state of memtx for the latest txn confirmed with quorum
- perform snapshot for the confirmed txn 
- reset WAL to be empty

**Comment by kostja** [comment]
There is no need to do it if RAFT is running strictly in WAL thread, the current machinery with wal_opt_rotate() will work just fine, except it will have to write a dummy quorum transaction to the RAFT log, not just make a round trip over cbus.

**Comment by sergos** [comment]
But if we will not obtain quorum on the txn we're trying to make a snapshot for? 
Then we will replay the undo and appear in an inconsistent state.

**Comment by kostja** [comment]
wal_opt_rotate() will ensure the snapshot is not "committed" if there is no quorum on the last transaction in the snapshot, since it flushes the cbus. So basically the snapshot procedure will stay as it is now:
1. Freeze database state, including the dirty (uncommitted) data for snapshot
2. Send wal_opt_rotate() to cbus
3. Write the frozen state to disk, use .snap.inprogress file name
4. If/when wal_opt_rotate() returns, which means dirty state is now committed, rename .snap.inprogress to .snap, this marks the end of a successful snapshot.
 If it doesn't, abandon .inprogress (it will be deleted by garbage collection) and return error.

**Comment by sergos** [comment]
So this is something I mentioned today as 'optimistic approach' when we speculatively assume confirmation, in contrast to what I described in ticket as 'pessimistic' one.
Ok, I won't object.
